File: plugin/com.hello.uxml.tools.ui/src/com/hello/uxml/tools/framework/GridColumn.java
package com.hello.uxml.tools.framework;

import java.util.EnumSet;

/**
 * Implements a grid column.
 *
 * @author ferhat
 */
public class GridColumn extends GridDef {

  /** Width Property Definition */
  public static PropertyDefinition widthPropDef = PropertySystem.register("Width", Double.class,
      GridColumn.class,
      new PropertyData(Double.NaN, EnumSet.of(PropertyFlags.None), new PropertyChangeListener() {
        @Override
        public void propertyChanged(PropertyChangedEvent e) {
          ((GridColumn) e.getSource()).length = ((Double) e.getNewValue()).doubleValue();
        }}));

  /** MinWidth Property Definition */
  public static PropertyDefinition minWidthPropDef = PropertySystem.register("MinWidth",
      Double.class, GridColumn.class,
      new PropertyData(Double.NaN, EnumSet.of(PropertyFlags.None), new PropertyChangeListener() {
        @Override
        public void propertyChanged(PropertyChangedEvent e) {
          ((GridColumn) e.getSource()).minLength = ((Double) e.getNewValue()).doubleValue();
        }}));

  /** MaxWidth Property Definition */
  public static PropertyDefinition maxWidthPropDef = PropertySystem.register("MaxWidth",
      Double.class, GridColumn.class,
      new PropertyData(Double.NaN, EnumSet.of(PropertyFlags.None), new PropertyChangeListener() {
        @Override
        public void propertyChanged(PropertyChangedEvent e) {
          ((GridColumn) e.getSource()).maxLength = ((Double) e.getNewValue()).doubleValue();
        }}));

  /**
   * Sets/returns width property.
   */
  public void setWidth(double value) {
    setProperty(widthPropDef, value);
  }

  public double getWidth() {
    return length;
  }

  /**
   * Sets/returns minimum width property.
   */
  public void setMinWidth(double value) {
    setProperty(minWidthPropDef, value);
  }

  public double getMinWidth() {
    return minLength;
  }

  /**
   * Sets/returns maximum width property.
   */
  public void setMaxWidth(double value) {
    setProperty(maxWidthPropDef, value);
  }

  public double getMaxWidth() {
    return maxLength;
  }
}

File: framework/lib/grid_column.dart
part of uxml;

/**
 * Implements a grid column.
 *
 * @author ferhat@ (Ferhat Buyukkokten)
 */
class GridColumn extends GridDef {
  static ElementDef gridcolumnElementDef;
  /**
   * MinWidth Property Definition.
   */
  static PropertyDefinition minWidthProperty;
  /**
   * MaxWidth Property Definition.
   */
  static PropertyDefinition maxWidthProperty;

  /**
   * Width Property Definition.
   */
  static PropertyDefinition widthProperty;

  GridColumn() : super() {
  }

  /**
   * Gets or sets width of a grid column.
   */
  num get width => _length;

  set width(num value) {
    setProperty(widthProperty, value);
  }

  /**
   * Gets or sets the minimum width of a grid column.
   */
  num get minWidth => minLength;

  set minWidth(num value) {
    setProperty(minWidthProperty, value);
  }

  /**
   * Gets or sets the maximum width of a grid column.
   */
  num get maxWidth => maxLength;

  set maxWidth(num value) {
    setProperty(maxWidthProperty, value);
  }

  static void _widthChangedHandler(UxmlElement target,
                                   PropertyDefinition property,
                                   Object oldValue,
                                   Object newValue) {
    GridColumn column = target;
    if (newValue == PropertyDefaults.NO_DEFAULT) {
      column.lengthDefined = false;
    } else {
      column._length = newValue;
      column.lengthDefined = true;
    }
    if (column.owner != null) {
      column.owner._gridDefChanged();
    }
  }

  static void _maxWidthChangedHandler(UxmlElement target,
                                      PropertyDefinition property,
                                      Object oldValue,
                                      Object newValue) {
    GridColumn column = target;
    if (newValue == PropertyDefaults.NO_DEFAULT) {
      column.maxLengthDefined = false;
    } else {
      column.maxLength = newValue;
    }
    if (column.owner != null) {
      column.owner._gridDefChanged();
    }
  }

  static void _minWidthChangedHandler(UxmlElement target,
                                      PropertyDefinition property,
                                      Object oldValue,
                                      Object newValue) {
    GridColumn column = target;
    column.minLength = (newValue == PropertyDefaults.NO_DEFAULT) ? 0 : newValue;
    if (column.owner != null) {
      column.owner._gridDefChanged();
    }
  }

  /** @see UxmlElement.getDefinition. */
  ElementDef getDefinition() => gridcolumnElementDef;

  /** Registers component. */
  static void registerGridColumn() {
    minWidthProperty = ElementRegistry.registerProperty("minWidth",
        PropertyType.NUMBER, PropertyFlags.NONE, _minWidthChangedHandler, 0.0);
    maxWidthProperty = ElementRegistry.registerProperty("maxWidth",
        PropertyType.NUMBER, PropertyFlags.NONE, _maxWidthChangedHandler, 0.0);
    widthProperty = ElementRegistry.registerProperty("width",
        PropertyType.NUMBER, PropertyFlags.NONE, _widthChangedHandler, 0.0);

    gridcolumnElementDef = ElementRegistry.register("GridColumn",
        UxmlElement.baseElementDef,
        [minWidthProperty, maxWidthProperty, widthProperty], null);
  }
}
